File: step_04_binary_search/IB031_search_in_bitonic_array/solution.py
"""Problem: Search in Bitonic Array
Difficulty: MEDIUM | XP: 25
Source: InterviewBit

A bitonic array first increases then decreases. Find the index
of a target element. Return -1 if not found.
Real-life use: searching in unimodal datasets (elevation profiles, profit curves).
"""
from typing import List

# ============================================================
# APPROACH 1: BRUTE FORCE
# Time: O(n)  |  Space: O(1)
# Linear scan
# ============================================================
def brute_force(A: List[int], target: int) -> int:
    for i, v in enumerate(A):
        if v == target:
            return i
    return -1


# ============================================================
# APPROACH 2: OPTIMAL
# Time: O(log n)  |  Space: O(1)
# 1. Binary search to find peak index
# 2. Binary search (ascending) in [0..peak]
# 3. Binary search (descending) in [peak+1..n-1]
# ============================================================
def optimal(A: List[int], target: int) -> int:
    n = len(A)

    # Find peak
    lo, hi = 0, n - 1
    while lo < hi:
        mid = (lo + hi) // 2
        if A[mid] < A[mid + 1]:
            lo = mid + 1
        else:
            hi = mid
    peak = lo

    # Ascending binary search
    lo, hi = 0, peak
    while lo <= hi:
        mid = (lo + hi) // 2
        if A[mid] == target:
            return mid
        elif A[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1

    # Descending binary search
    lo, hi = peak + 1, n - 1
    while lo <= hi:
        mid = (lo + hi) // 2
        if A[mid] == target:
            return mid
        elif A[mid] > target:
            lo = mid + 1
        else:
            hi = mid - 1

    return -1


# ============================================================
# APPROACH 3: BEST
# Time: O(log n)  |  Space: O(1)
# Same 3-phase approach, refactored into helper functions
# for cleaner structure.
# ============================================================
def best(A: List[int], target: int) -> int:
    def find_peak(arr):
        lo, hi = 0, len(arr) - 1
        while lo < hi:
            mid = (lo + hi) // 2
            if arr[mid] < arr[mid + 1]:
                lo = mid + 1
            else:
                hi = mid
        return lo

    def bs_asc(arr, lo, hi, t):
        while lo <= hi:
            mid = (lo + hi) // 2
            if arr[mid] == t: return mid
            if arr[mid] < t: lo = mid + 1
            else: hi = mid - 1
        return -1

    def bs_desc(arr, lo, hi, t):
        while lo <= hi:
            mid = (lo + hi) // 2
            if arr[mid] == t: return mid
            if arr[mid] > t: lo = mid + 1
            else: hi = mid - 1
        return -1

    peak = find_peak(A)
    res = bs_asc(A, 0, peak, target)
    return res if res != -1 else bs_desc(A, peak + 1, len(A) - 1, target)


if __name__ == "__main__":
    cases = [
        ([1, 3, 8, 12, 4, 2], 4, 4),
        ([3, 9, 10, 20, 17, 5, 1], 20, 3),
        ([5, 6, 7, 8, 9, 10, 3, 2, 1], 3, 6),
        ([10, 20, 30, 40, 50], 5, -1),
        ([1, 2, 3], 1, 0),
    ]
    print("=== Search in Bitonic Array ===")
    for A, target, exp in cases:
        b   = brute_force(A[:], target)
        o   = optimal(A[:], target)
        bst = best(A[:], target)
        ok = "OK" if b == o == bst == exp else f"MISMATCH(BF={b},OPT={o},BEST={bst})"
        print(f"A={A}  target={target} => {ok}  EXP={exp}")

File: step_04_binary_search/IB031_search_in_bitonic_array/Solution.java
import java.util.*;

// ============================================================
// APPROACH 1: BRUTE FORCE
// Time: O(n)  |  Space: O(1)
// Linear scan through the entire array.
// ============================================================
class BruteForce {
    public static int solve(int[] A, int target) {
        for (int i = 0; i < A.length; i++)
            if (A[i] == target) return i;
        return -1;
    }
}

// ============================================================
// APPROACH 2: OPTIMAL
// Time: O(log n)  |  Space: O(1)
// 1. Find the peak (bitonic point) using binary search.
// 2. Binary search in ascending half [0..peak].
// 3. Binary search in descending half [peak+1..n-1].
// ============================================================
class Optimal {
    private static int findPeak(int[] A) {
        int lo = 0, hi = A.length - 1;
        while (lo < hi) {
            int mid = (lo + hi) / 2;
            if (A[mid] < A[mid + 1]) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }

    private static int bsAsc(int[] A, int lo, int hi, int target) {
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            if (A[mid] == target) return mid;
            if (A[mid] < target) lo = mid + 1;
            else hi = mid - 1;
        }
        return -1;
    }

    private static int bsDesc(int[] A, int lo, int hi, int target) {
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            if (A[mid] == target) return mid;
            if (A[mid] > target) lo = mid + 1;
            else hi = mid - 1;
        }
        return -1;
    }

    public static int solve(int[] A, int target) {
        int peak = findPeak(A);
        int idx = bsAsc(A, 0, peak, target);
        return idx != -1 ? idx : bsDesc(A, peak + 1, A.length - 1, target);
    }
}

// ============================================================
// APPROACH 3: BEST
// Time: O(log n)  |  Space: O(1)
// Same as Optimal, written more concisely and combined in
// a single unified method with clear structure.
// ============================================================
class Best {
    public static int solve(int[] A, int target) {
        int n = A.length;
        // Find peak index
        int lo = 0, hi = n - 1;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (A[mid] < A[mid + 1]) lo = mid + 1;
            else hi = mid;
        }
        int peak = lo;

        // Search ascending part [0..peak]
        lo = 0; hi = peak;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (A[mid] == target) return mid;
            if (A[mid] < target) lo = mid + 1; else hi = mid - 1;
        }

        // Search descending part [peak+1..n-1]
        lo = peak + 1; hi = n - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (A[mid] == target) return mid;
            if (A[mid] > target) lo = mid + 1; else hi = mid - 1;
        }

        return -1;
    }
}

public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Search in Bitonic Array ===");
        int[][] arrays = {
            {1, 3, 8, 12, 4, 2},
            {3, 9, 10, 20, 17, 5, 1},
            {5, 6, 7, 8, 9, 10, 3, 2, 1},
            {10, 20, 30, 40, 50},
        };
        int[] targets = {4, 20, 3, 5};
        int[] expected = {4, 3, 6, -1};

        for (int t = 0; t < arrays.length; t++) {
            int bf  = BruteForce.solve(arrays[t], targets[t]);
            int op  = Optimal.solve(arrays[t], targets[t]);
            int bst = Best.solve(arrays[t], targets[t]);
            String ok = (bf == op && op == bst && bst == expected[t]) ? "OK" : "MISMATCH";
            System.out.println("arr=" + Arrays.toString(arrays[t]) + " target=" + targets[t]
                + "  BF=" + bf + "  OPT=" + op + "  BEST=" + bst + "  EXP=" + expected[t] + "  " + ok);
        }
    }
}
